proof_of_play.go

```diff
@@ -158,3 +172,40 @@ func (p *proofOfPlay) expire(ad Ad) error {
 		return false, nil
 	})
 }
+
+func (p *proofOfPlay) isLeaseExpired(ad Ad) bool {
+	exp, ok := ad["lease_expiry"].(float64)
+	if !ok {
+		return true
+	}
+
+	expiry := time.Unix(int64(exp), 0)
+	return time.Now().After(expiry)
+}
+
+func (p *proofOfPlay) processRetries() {
+	for req := range p.retryQueue {
+		p.retryPoP(req)
+	}
+}
+
+func (p *proofOfPlay) retryPoP(req *PoPRequest) {
+	ad := req.Ad
+
+	if p.isLeaseExpired(ad) {
+		popType := "pop"
+		if !req.Status {
+			popType = "expire"
+		}
+
+		p.publishEvent(
+			fmt.Sprintf("ad-%s-already-expired", popType),
+			fmt.Sprintf("ad: %s, expiry: %d", ad["id"], ad["lease_expiry"]),
+			"critical")
+		return
+	}
+
+	sleepDuration := RetryInterval - time.Since(req.RequestTime)
```

Comment:
This will work if retryQueue elements have RequestTime's strictly in increasing order. Otherwise, you may end up with a negative sleep duration. I don't think your current implementation guarantees this ordering because you're setting RequestTime before making the expire / confirm calls. You may end up out of sync elements if expire / confirms take different times to complete. Think about this:

1. req1 arrives
2. req1.RequestTime is set to 10
3. req2 arrives
4. req2.RequestTime is set to 11
5. confirm(req1) gets called
6. confirm(req2) gets called
7. confirm(req2) fails at 15 and added to retry queue
8. confirm(req1) fails at 16 and added to retry queue

At this point you'll end up [req2, req1] in the retry queue with request times of [11, 10].